server/resp/src/test/java/org/infinispan/server/resp/commands/list/blocking/PollListenerSynchronizerTest.java

```diff
@@ -0,0 +1,223 @@
+   public void raceEventAndPoll(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      var l = new CountDownLatch(1);
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onEvent("key1".getBytes());
+      }).start();
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      }).start();
+      l.countDown();
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void raceEventsAndPoll(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      var l = new CountDownLatch(1);
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onEvent("key1".getBytes());
+      }).start();
+
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onEvent("key2".getBytes());
+      }).start();
+
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      }).start();
+      l.countDown();
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void concurrentEventAndPollGotNullResult(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      new Thread(() -> {
+         try {
+            synchronizer.pollGotNullResultLatch.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onEvent("key1".getBytes());
+      }).start();
+      new Thread(() -> {
+         synchronizer.onPollComplete(null);
+      }).start();
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("key1".getBytes(), "test-val".getBytes()));
+
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void concurrentNullPollAndEventReceived(BLPOP.PollListenerSynchronizer synchronizer)
```

Comment:
With this synchronization here, I think this should be equivalent to the `onEventThenNullPollTest` method.